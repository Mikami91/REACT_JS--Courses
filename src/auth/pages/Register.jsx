import { useSelector } from 'react-redux';
import { useMemo, useState } from 'react';
import { Link as RouterLink } from 'react-router-dom';
import {
  Alert,
  Button,
  Grid,
  Link,
  TextField,
  Typography,
} from '@mui/material';
// Layouts
import { FormLayout } from '../layout';
// Hooks
import { useForm } from '../../hooks';
import { useDispatchAuth } from '../../store/auth/hooks';

const formData = {
  displayName: 'Nameless',
  email: 'nameless@gmail.com',
  password: '123456',
};

const errorMessages = [
  'The email must have a @',
  'The password must have 6 or more characters',
  'The username must have 4 or more characters',
];

const formValidations = {
  displayName: [(value) => value.length >= 4, errorMessages[2]],
  email: [(value) => value.includes('@'), errorMessages[0]],
  password: [(value) => value.length >= 6, errorMessages[1]],
};

const Register = () => {
  const [formSubmitted, setFormSubmitted] = useState(false);
  const { checkingAuthentication, startEmailSingIn } = useDispatchAuth();
  const { status, success, errorMessage } = useSelector((state) => state.auth);
  const {
    displayName,
    email,
    password,
    onInputChange,
    formState,
    displayNameValid,
    emailValid,
    passwordValid,
    isFormValid,
  } = useForm(formData, formValidations);

  const isAuthtenticated = useMemo(() => status === 'checking', [status]);

  const onSubmit = (e) => {
    e.preventDefault();
    setFormSubmitted(true);

    if (!isFormValid) return;

    checkingAuthentication();

    startEmailSingIn(formState);
  };

  return (
    <>
      <FormLayout
        title='Register'
        children={
          <form
            className='animate__animated animate__fadeIn animate__faster'
            onSubmit={onSubmit}
          >
            <p>Form valid: {isFormValid ? 'Correct' : 'Incorrect'}</p>
            <Grid container spacing={2} mt={2}>
              <Grid item xs={12}>
                <TextField
                  label='Username'
                  type='text'
                  placeholder='Your name'
                  autoComplete='new-password'
                  name='displayName'
                  value={displayName}
                  onChange={onInputChange}
                  error={!!displayNameValid && formSubmitted}
                  helperText={formSubmitted ? displayNameValid : ''}
                  fullWidth
                />
              </Grid>
              <Grid item xs={12}>
                <TextField
                  label='Email'
                  type='email'
                  placeholder='youremail@gmail.com'
                  autoComplete='new-password'
                  name='email'
                  value={email}
                  onChange={onInputChange}
                  error={!!emailValid && formSubmitted}
                  helperText={formSubmitted ? emailValid || errorMessage : ''}
                  fullWidth
                />
              </Grid>
              <Grid item xs={12}>
                <TextField
                  label='Password'
                  type='password'
                  placeholder='Password'
                  inputProps={{
                    autoComplete: 'new-password',
                    form: {
                      autoComplete: 'off',
                    },
                  }}
                  name='password'
                  value={password}
                  onChange={onInputChange}
                  error={!!passwordValid && formSubmitted}
                  helperText={formSubmitted ? passwordValid : ''}
                  fullWidth
                />
              </Grid>
            </Grid>
            <Grid container spacing={2} mt={2}>
              <Grid item xs={12} display={!!errorMessage ? '' : 'none'}>
                <Alert severity='error'>{errorMessage}</Alert>
              </Grid>
              <Grid item xs={12}>
                <Button
                  variant='contained'
                  type='submit'
                  disabled={isAuthtenticated}
                  // disabled={!isFormValid || formSubmitted}
                  fullWidth
                >
                  Register
                </Button>
              </Grid>

              <Grid container direction='row' justifyContent='end' mt={2}>
                <Typography mr={1}>Do you already have an account?</Typography>
                <Link component={RouterLink} color='inherit' to='/auth/login'>
                  Login
                </Link>
              </Grid>
            </Grid>
          </form>
        }
      />
    </>
  );
};

export default Register;
