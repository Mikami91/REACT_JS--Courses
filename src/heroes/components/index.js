export { default as HeroCard } from './HeroCard';
export { default as HeroResult } from './HeroResult';
export { default as HeroList } from './HeroList';
