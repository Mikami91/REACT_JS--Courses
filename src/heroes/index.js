export * from './pages';
export * from './routers/HeroesRouter';
