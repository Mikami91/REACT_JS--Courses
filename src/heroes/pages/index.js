export { default as DC } from './DC';
export { default as Hero } from './Hero';
export { default as Marvel } from './Marvel';
export { default as Publisher } from './Publisher';
export { default as Search } from './Search';
