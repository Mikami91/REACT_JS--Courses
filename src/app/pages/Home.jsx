// Components
import { FloatButton } from '../components';
// Layouts
import JournalLayout from '../layout/JournalLayout';
// Views
import { NoteView, NothingSelectedView } from '../views';

const Home = () => {
  return (
    <>
      <JournalLayout>
        {/* <NothingSelectedView /> */}
        <NoteView />
        <FloatButton />
      </JournalLayout>
    </>
  );
};

export default Home;
