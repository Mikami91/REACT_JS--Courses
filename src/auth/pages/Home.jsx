// Components
import { PageTitle } from '../../ui/components';

const Home = () => {
  return (
    <>
      <PageTitle title='Home Page' />
    </>
  );
};

export default Home;
