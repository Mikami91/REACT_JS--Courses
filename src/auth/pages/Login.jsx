import { useContext, useRef } from 'react';
import { useNavigate } from 'react-router-dom';
// Components
import { PageTitle } from '../../ui/components';
// Context
import { AuthContext } from '../context/';
// Hooks
import { useScreen } from '../../ui/hooks';

const Login = () => {
  const { login } = useContext(AuthContext);
  const { height } = useScreen();
  const navigate = useNavigate();
  const inputRef = useRef();

  const onLogin = (e) => {
    e.preventDefault();

    const username = inputRef.current.value || '';

    if (username.length < 2 || username.length > 20) return;

    const lastPath = localStorage.getItem('lastPath') || 'publisher';

    login(username.trim());

    navigate(lastPath, { replace: true });
  };

  return (
    <>
      <PageTitle title='Login Page' />

      <div className='container text-center'>
        <div
          style={{
            marginTop: `${height / 3.5}px`,
            marginBottom: `${height / 3.5}px`,
          }}
        >
          <form onSubmit={onLogin}>
            <div className='row justify-content-center mb-4'>
              <div className='col-xs-12 col-md-5 col-lg-3'>
                <input
                  type='text'
                  placeholder='Username'
                  className='form-control'
                  autoComplete='off'
                  ref={inputRef}
                />
              </div>
            </div>
            <div className='row justify-content-center'>
              <div className='col-xs-12 col-md-5 col-lg-3'>
                <button
                  className='form-control btn btn-outline-success'
                  type='button'
                  onClick={onLogin}
                >
                  Login
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </>
  );
};

export default Login;
