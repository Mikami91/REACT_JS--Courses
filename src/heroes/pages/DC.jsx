// Components
import { HeroList } from '../components';
import { PageTitle } from '../../ui/components';

const DC = () => {
  return (
    <>
      <PageTitle title='DC Page' />
      <HeroList publisher={'DC Comics'} />
    </>
  );
};

export default DC;
