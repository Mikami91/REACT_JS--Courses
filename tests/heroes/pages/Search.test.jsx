import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
// Pages
import { Search } from '../../../src/heroes/pages';

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}));

describe('TEST: <Search />', () => {
  beforeEach(() => jest.clearAllMocks());

  test('Should show with correct components', () => {
    const { constainer } = render(
      <MemoryRouter initialEntries={['/publisher/search']}>
        <Search />
      </MemoryRouter>
    );

    screen.debug();

    expect(constainer).toMatchSnapshot();
  });

  test('Should show hero info', () => {
    render(
      <MemoryRouter initialEntries={['/publisher/search?q=batman']}>
        <Search />
      </MemoryRouter>
    );

    const input = screen.getByRole('textbox');
    const img = screen.getByRole('img');

    expect(input.value).toBe('batman');
    expect(img.src).toContain('/heroes/dc-batman.jpg');

    screen.debug();
  });

  test('Should show error if not find a hero', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/publisher/search?q=mikami']}>
        <Search />
      </MemoryRouter>
    );

    const child = container.getElementsByClassName('alert-danger');

    expect(child.length).toBe(1);
  });

  test('Should call the navigate to hero page', () => {
    render(
      <MemoryRouter initialEntries={['/publisher/search']}>
        <Search />
      </MemoryRouter>
    );

    const inputValue = 'spiderman';

    const input = screen.getByRole('textbox');

    fireEvent.change(input, {
      target: { name: 'searchText', value: inputValue },
    });

    fireEvent.submit(input);

    expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);
  });
});
