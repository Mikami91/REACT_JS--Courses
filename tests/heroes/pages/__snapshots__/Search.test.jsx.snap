// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST: <Search /> Should show with correct components 1`] = `undefined`;
